eth/connmanager.go

```diff
@@ -0,0 +1,254 @@
+type getPeersFunc func() []*p2p.Peer
+
+// Callback type to get syncing status.
+// Returns true while syncing, false when synced.
+type getSyncingFunc func() bool
+
+type connmanConfig struct {
+	maxDialPeers    int // maximum number of dialed peers
+	maxInboundPeers int // maximum number of inbound peers
+	log             log.Logger
+	clock           mclock.Clock
+	rand            *mrand.Rand
+}
+
+func (cfg connmanConfig) withDefaults() connmanConfig {
+	if cfg.log == nil {
+		cfg.log = log.Root()
+	}
+	if cfg.clock == nil {
+		cfg.clock = mclock.System{}
+	}
+	if cfg.rand == nil {
+		seedb := make([]byte, 8)
+		crand.Read(seedb)
+		seed := int64(binary.BigEndian.Uint64(seedb))
+		cfg.rand = mrand.New(mrand.NewSource(seed))
+	}
+	return cfg
+}
+
+func newConnManager(config *connmanConfig) *connManager {
+	cfg := config.withDefaults()
+	cm := &connManager{
+		connmanConfig:        cfg,
+		peerDropDialedTimer:  mclock.NewAlarm(cfg.clock),
+		peerDropInboundTimer: mclock.NewAlarm(cfg.clock),
+		peerEventCh:          make(chan *p2p.PeerEvent),
+		shutdownCh:           make(chan struct{}),
+	}
+	if peerDropIntervalMin > peerDropIntervalMax {
+		panic("peerDropIntervalMin duration must be less than or equal to peerDropIntervalMax duration")
+	}
+	cm.log.Info("New Connection Manager", "maxDialPeers", cm.maxDialPeers, "threshold", peerDropThreshold,
+		"intervalMin", peerDropIntervalMin, "intervalMax", peerDropIntervalMax)
+	return cm
+}
+
+// Start the connection manager.
+func (cm *connManager) Start(srv *p2p.Server, syncingFunc getSyncingFunc) {
+	cm.wg.Add(1)
+	cm.peersFunc = srv.Peers
+	cm.syncingFunc = syncingFunc
+	cm.sub = srv.SubscribeEvents(cm.peerEventCh)
+	go cm.loop()
+}
+
+// Stop the connection manager.
+func (cm *connManager) Stop() {
+	cm.sub.Unsubscribe()
+	cm.peerDropInboundTimer.Stop()
+	cm.peerDropDialedTimer.Stop()
+	close(cm.shutdownCh)
+	cm.wg.Wait()
+}
+
+// numPeers returns the current number of peers and its breakdown (dialed or inbound).
+func (cm *connManager) numPeers() (numPeers int, numDialed int, numInbound int) {
+	peers := cm.peersFunc()
+	dialed := slices.DeleteFunc(peers, (*p2p.Peer).Inbound)
+	return len(peers), len(dialed), len(peers) - len(dialed)
+}
+
+// dropRandomPeer selects one of the peers randomly and drops it from the peer pool.
+func (cm *connManager) dropRandomPeer(dialed bool) bool {
+	peers := cm.peersFunc()
+
+	selectDoNotDrop := func(p *p2p.Peer) bool {
+		if dialed {
+			// Only drop from dyndialed peers. Avoid dropping trusted peers.
+			// Give some time to peers before considering them for a drop.
+			return !p.DynDialed() ||
+				p.Trusted() ||
+				p.Lifetime() < mclock.AbsTime(doNotDropBefore)
+		} else {
+			// Only drop from inbound peers. Avoid dropping trusted peers.
+			// Give some time to peers before considering them for a drop.
+			return p.DynDialed() || p.StaticDialed() ||
+				p.Trusted() ||
+				p.Lifetime() < mclock.AbsTime(doNotDropBefore)
```

Comment:
whole condition just simplifies to `p.StaticDialed() || p.Trusted() || p.Lifetime() < ...`